cpu/atmega_common/periph/i2c.c

```diff
@@ -0,0 +1,347 @@
+ *
+ * @author      Dimitri Nahm <dimitri.nahm@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+
+#include "cpu.h"
+#include "mutex.h"
+#include "periph/i2c.h"
+#include "periph_conf.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define MT_START 0x08
+#define MT_ADDRESS_ACK 0x18
+#define MT_DATA_ACK 0x28
+#define MR_ADDRESS_ACK 0x40
+
+/* guard file in case no I2C device is defined */
+#if I2C_NUMOF
+
+/* static function definitions */
+static int _start(uint8_t address, uint8_t rw_flag);
+static int _write(const uint8_t *data, int length);
+static void _stop(void);
+
+/**
+ * @brief Array holding one pre-initialized mutex for each I2C device
+ */
+static mutex_t locks[] =  {
+#if I2C_0_EN
+    [I2C_0] = MUTEX_INIT,
+#endif
+};
+
+int i2c_init_master(i2c_t dev, i2c_speed_t speed)
+{
+    /* TWI Bit Rate Register - division factor for the bit rate generator*/
+    int twibrr;
+
+    /* check if the line is valid */
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* calculate speed configuration */
+    switch (speed) {
+        case I2C_SPEED_NORMAL:
+            twibrr = ((CLOCK_CORECLOCK/100000)-16)/2;
+            break;
+
+        case I2C_SPEED_FAST:
+            twibrr = ((CLOCK_CORECLOCK/400000)-16)/2;
+            break;
+
+        default:
+            return -2;
+    }
+
+    /* set pull-up on SCL and SDA */
+    #if defined (CPU_ATMEGA2560) || defined (CPU_ATMEGA1281)
+        PORTD |= (1 << PORTD0) | (1 << PORTD1);
+    #endif
+    #ifdef CPU_ATMEGA328P
+        PORTC |= (1 << PORTC4) | (1 << PORTC5);
+    #endif
+
+    /* enable I2C clock */
+    i2c_poweron(dev);
+
+    /* disable device */
+    TWCR &= ~(1 << TWEN);
+    /* configure I2C clock */
+    TWBR = twibrr;
+    /* enable device */
+    TWCR |= (1 << TWEN);
+
+    return 0;
+}
+
+int i2c_acquire(i2c_t dev)
+{
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+    mutex_lock(&locks[dev]);
+    return 0;
+}
+
+int i2c_release(i2c_t dev)
+{
+    if (dev >= I2C_NUMOF) {
+        return -1;
+    }
+    mutex_unlock(&locks[dev]);
+    return 0;
+}
+
+int i2c_read_byte(i2c_t dev, uint8_t address, void *data)
+{
+    return i2c_read_bytes(dev, address, data, 1);
+}
+
+int i2c_read_bytes(i2c_t dev, uint8_t address, void *data, int length)
+{
+    uint8_t *my_data = data;
+
+    if ((unsigned int)dev >= I2C_NUMOF) {
+        return -1;
+    }
+
+    /* send start condition and slave address */
+    if (_start(address, I2C_FLAG_READ) != 0)
+        return 0;
+
+    for (int i = 0; i < length; i++) {
+        /* Send NACK for last received byte */
+        if ((length-i) == 1)
+            TWCR = (1 << TWEN) | (1 << TWINT);
+        else
+            TWCR = (1 << TWEA) | (1 << TWEN) | (1 << TWINT);
+        DEBUG("Wait for byte %i\n", i+1);
+        /* Wait for TWINT Flag set. This indicates that DATA has been received.*/
+        while (!(TWCR & (1 << TWINT)))
+            {}
```

Comment:
I think the common style is to keep the `{}` in the same line as the `while` statement